WordPress/gutenberg
# Forbid using certain functions and classes

## What?
This PR introduces the `ForbiddenFunctionsAndClasses` sniff.
It's designed to enforce usage restrictions on specific classes and functions within the Gutenberg project.
While declarations of these classes/functions are permitted, their actual usage is restricted based on defined regular expressions.

Fixes https://github.com/WordPress/gutenberg/issues/54745.

## Why?
The necessity of this PR arises from the need to avoid using any Gutenberg functions and classes in the block-library packages.
This is because these packages are backported to Core and must not include Gutenberg related functions and classes.

## How?
The implementation involves defining regular expressions for restricted classes and functions.
The `ForbiddenFunctionsAndClassesSniff` class hooks into the PHP CodeSniffer process, scanning the code for instances where these restricted classes or functions are used.


## Testing Instructions
1. Ensure that GitHub CI jobs pass.
2. Navigate to `test/php/gutenberg-coding-standards/`.
3. Run `composer update`.
4. Run `composer run check-all`.

### Testing Instructions for Keyboard


## Screenshots or screencast
